### PR TITLE
Improve performance of pod logs

### LINKED ISSUE #7156: Reports of browsers running out of memory when tailing log files in UI
Internal reference: SURE-5383
Reported in Rancher 2.6.8.

When troubleshooting an issue involving websockets, I did get a report that was adjacent to it. When using the Vue UI to tail log files, the browser stops responding.

> browser memory is consumed until the tab is killed. If you leave one open long enough, it will eventually die. If you pull up pod logs for a busy pod though you can kill it in a few minutes (rancher kubectl UI). We have tried, edge, chrome and brave and they all exhibit the symptom.

> Browser tab started at 650mb before opening logs, high cpu usage while streaming them, and memory growing rapidly. Browser tab crashed at about 2.5gb memory footprint.

We'll need to see if we can repro this to narrow down what's going on. We may need busy log activity to fully reproduce.

Workaround:
Restart browser tab every few minutes, or sometimes before one minute.

### PR DESCRIPTION
**Pull request by mantis-toboggan-md**
### Summary
Fixes #7156 

### Occurred changes and/or fixed issues
This PR updates the `ContainerLogs` component to use a virtual list- only the log lines visible on the page (plus a few more for padding) are rendered. The component will also only store the last x lines when that range option is selected, eg if the user has selected "show the last 10,000 lines" the containerlogs component only retains the latest 10,000 lines - previously we retrieved the last 10K then continuously appended new logs. 

### Areas or cases that should be tested
I used chrome's performance monitor to test this, which you can find in the chrome devtool's kebab menu:
Screen Shot 2022-11-21 at 9 26 36 AM

* Containers with various log activity should be tested. I used a deployment with 3 containers each logging something in a different interval (pasted below) 
* Various log range options 
* Resizing and moving the container between the bottom and size of the page
* Follow button/scrolling un-follow behavior 
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    workload.user.cattle.io/workloadselector: apps.deployment-default-test-logs
  name: test-logs
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      workload.user.cattle.io/workloadselector: apps.deployment-default-test-logs
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      labels:
        workload.user.cattle.io/workloadselector: apps.deployment-default-test-logs
    spec:
      affinity: {}
      containers:
      - args:
        - 'i=0; while true; do echo "$i: $(date)"; i=$((i+1)); sleep .001; done'
        command:
        - /bin/sh
        - -c
        image: busybox
        imagePullPolicy: Always
        name: fast
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      - args:
        - 'i=0; while true; do echo "$i: $(date)"; i=$((i+1)); sleep 1; done'
        command:
        - /bin/sh
        - -c
        image: busybox
        imagePullPolicy: Always
        name: slower
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      - args:
        - 'i=0; while true; do echo "$i: $(date)"; i=$((i+1)); sleep 10; done'
        command:
        - /bin/sh
        - -c
        image: busybox
        imagePullPolicy: Always
        name: sloooow
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
```

### CODE REVIEW

File: shell/components/nav/WindowManager/ContainerLogs.vue [hunk 1]
```diff
@@ -592,44 +583,17 @@ export default {
...
-              class="msg text-muted"
-            />
-          </tbody>
-        </table>
+        <VirtualList
+          v-show="filtered.length"
+          ref="virtualList"
+          data-key="id"
+          :data-sources="filtered"
+          :data-component="logItem"
```
Comment (rak-phillip) [inline shell/components/nav/WindowManager/ContainerLogs.vue:591]: NOTE: I found this usage of `data-component` quite interesting.. I can't say that I've seen passing a component as a prop before, but I suppose this is an implementation decision of how we are expected to interact with the new library.

**Review by rak-phillip — CHANGES_REQUESTED** [review]
Functionality looks great. Left some feedback based on some observations that I made while reviewing.

## Time Format

It looks like the format for timestamps is different between the original and PR 7511. Is this difference intentional?

### Original

image

### PR

image

## Empty Message

It looks like we lose the empty message in this PR.

### Original

image

### PR

image

## Typeface

The typeface looks different in this change

### Original

image

### PR

image

File: shell/components/LogItem.vue [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+      return day(time).format(this.timeFormatStr);
+    }
+  }
+};
+</script>
+
+<template>
+  <div>
+    <span v-html="format(source.time)" />
```
[the remark this answers is not in the document]
Comment (mantis-toboggan-md) [reply shell/components/LogItem.vue]: We don't expect html for the time, so I've updated that, but we can have raw html in the message: we're converting color codes to html in the socket message event handler.

File: shell/components/LogItem.vue [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    }
+  },
+
+  methods: {
+    format(time) {
+      if ( !time ) {
+        return '';
+      }
+
+      return day(time).format(this.timeFormatStr);
```
[the remark this answers is not in the document]
Comment (mantis-toboggan-md) [reply shell/components/LogItem.vue:29]: It should be coming from date and time format user prefs; corrected now.

File: shell/components/nav/WindowManager/ContainerLogs.vue [hunk 4]
```diff
@@ -214,24 +222,26 @@ export default {
       const timeFormat = escapeHtml( this.$store.getters['prefs/get'](TIME_FORMAT));
 
       return `${ dateFormat } ${ timeFormat }`;
+    },
+
+    maxLines() {
```
[the remark this answers is not in the document]
Comment (mantis-toboggan-md) [reply shell/components/nav/WindowManager/ContainerLogs.vue]: Nope, I refactored this into redundancy. Thanks for catching that.

File: shell/components/nav/WindowManager/ContainerLogs.vue [hunk 5]
```diff
@@ -337,24 +320,32 @@ export default {
...
         this.backlog = [];
+        const maxLines = this.parseRange(this.range)?.tailLines;
+
+        if (maxLines && this.lines.length > maxLines) {
+          this.lines = this.lines.slice(-maxLines);
+        }
       }
 
-      if ( this.isFollowing ) {
+      if ( this.isFollowing) {
```
[the remark this answers is not in the document]
Comment (mantis-toboggan-md) [reply shell/components/nav/WindowManager/ContainerLogs.vue]: Fixed

File: shell/components/nav/WindowManager/ContainerLogs.vue (hunk 1 above)
Comment (mantis-toboggan-md) [reply shell/components/nav/WindowManager/ContainerLogs.vue:569]: Yeah I thought it was a little strange, but the library doesn't offer an alternative.

File: shell/components/nav/WindowManager/ContainerLogs.vue [hunk 6]
```diff
@@ -347,14 +321,17 @@ export default {
     },
 
     updateFollowing() {
-      const el = this.$refs.body;
+      const virtualList = this.$refs.virtualList;
 
-      this.isFollowing = el.scrollTop + el.clientHeight + 2 >= el.scrollHeight;
+      if (virtualList) {
+        this.isFollowing = virtualList.getScrollSize() - virtualList.getClientSize() === virtualList.getOffset();
```
Comment (rak-phillip) [inline shell/components/nav/WindowManager/ContainerLogs.vue:327]: Nitpicking here 🙂 

I think grouping with parentheses makes the expression easier to read 

```suggestion
        this.isFollowing = (virtualList.getScrollSize() - virtualList.getClientSize()) === virtualList.getOffset();
```

**Review by rak-phillip — APPROVED** [review]
LGTM

I left an optional comment related to code style.

**Review by rak-phillip — APPROVED** [review]
LGTM

**Pull request merged**